### PR TITLE
zchaff: 2004.5.13 -> 2007.3.12, fix coqPackages.smtcoq

### PR DESCRIPTION
**Pull request by siraben**
###### Description of changes
Update zchaff and fixes smtcoq. See https://github.com/smtcoq/smtcoq/issues/106 for details.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vbgl** [comment]
What’s the issue with `cvc4`?

**Comment by siraben** [comment]
https://github.com/smtcoq/smtcoq/issues/107

**Comment by siraben** [comment]
The `buildInputs`, `nativeBuildInputs` and cmake flags differed enough that copying the definition wholesale would be more maintainable.

File: pkgs/development/coq-modules/smtcoq/cvc4.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    url = "https://cvc4.cs.stanford.edu/downloads/builds/src/cvc4-${version}.tar.gz";
+    sha256 = "1iw793zsi48q91lxpf8xl8lnvv0jsj4whdad79rakywkm1gbs62w";
+  };
+
+  nativeBuildInputs = [ autoreconfHook pkg-config ];
+  buildInputs = [ gmp readline swig libantlr3c antlr3_4 boost jdk8 python3 ]
+    ++ lib.optionals stdenv.isLinux [ cln ];
+
+  configureFlags = [
+    "--enable-language-bindings=c,c++,java"
```
Comment (vbgl) [inline pkgs/development/coq-modules/smtcoq/cvc4.nix:20]: Are the java binding needed? IIUC they add a dependency to `jdk8`.

**Comment by vbgl** [comment]
@GrahamcOfBorg build coqPackages_8_13.smtcoq

**Comment by vbgl** [comment]
It seems that `smtcoq` is still broken…

**Comment by siraben** [comment]
@vbgl what error do you get?

**Comment by vbgl** [comment]
Full log: https://logs.nix.ci/?key=nixos/nixpkgs.176321&attempt_id=8e4aeff7-185e-4627-935c-57dc6bc02ce9

>   254 |     max(const _Tp& __a, const _Tp& __b)
      |     ^~~
/nix/store/zs6hnk47hcrrcrd885d66xmhlvp05jxm-gcc-11.3.0/include/c++/11.3.0/bits/stl_algobase.h:254:5: note:   template argument deduction/substitution failed:
../../../../../src/util/regexp.cpp:422:42: note:   candidate expects 2 arguments, 0 provided

**Comment by Zimmi48** [comment]
@siraben: Are you working on a fix? Just FYI, this is blocking the nixpkgs update in https://github.com/coq-community/coq-nix-toolbox/pull/114.

**Comment by siraben** [comment]
I am unable to replicate the issue on macOS. Let me know which platforms this is happening on.

**Comment by vbgl** [comment]
Follow the link I’ve put above: linux x86_64

**Comment by proux01** [comment]
@vbgl should we revert this until it gets fixed?

**Comment by vbgl** [comment]
Proposed fix in #183084.

**Pull request merged**